My Attempt to make the ORM Weld to the Floor

RedBaronFlyer opened the pull request:
## About The Pull Request
I was always a bit surprised about how the ORM could be unwrenched to move, as every other machine requires a multistep process to move. This, combined with most maps having the ORM be in the wall leads to maps such as icebox, tram, birdshot, and especially metastation being a wrench away from being tided into.

After experiencing a round where cargo got tided into nearly ten times (this was on Manuel, MRP moment) I decided to sit down and bother to get this working. This is basically my attempt to get [this PR](https://github.com/tgstation/tgstation/pull/70597) merged, since it went stale.

## Demonstration (I wasn't able to get this to fit down into a size of 10 MBs or less, sorry):
https://www.youtube.com/watch?v=k5rWaAinFxs

## Why It's Good For The Game

1. Cargo will now be as secure as other departments (such as botany) that have a similar setup with a machine in the wall by the door.

2. It will bring the ORM in line with other machines. Nearly every other machine in the game currently requires either:
A. Welding and then wrenching (Chef/Botanist/Chemist/Coronor fridges)
B. Deconstructing the entire machine (autolathe, department lathe, microwaves, etc.)

## Changelog

The ORM now requires welding it prior to unwrenching it, bringing it in line with other similar machines such as the chef's fridge vendor.

:cl:
balance: The ORM now requires welding it prior to wrenching it, similar to Chef/Botanist/Chemist/Coronor fridges.
/:cl:

RedBaronFlyer committed:
Changes the ORM to require welding before you can unwrench it

RedBaronFlyer committed:
Fixes a part of a line on 381 that was deleted (somehow)

RedBaronFlyer commented:
I'm keeping this as a draft at the moment just incase anything weird pops up. I somehow deleted part of line 381 prior to uploading this so I'm trying to be safe.

I didn't know what tag would fit best so I put balance.

mc-oofert commented:
how does this improve security of said department theres like 20 welding tools free on the map roundstart

RedBaronFlyer commented:
> how does this improve security of said department theres like 20 welding tools free on the map roundstart

It gives cargo a bigger window to stop someone from trying to tide as opposed to how it is right now where you can unwrench the ORM in less than three seconds.

GuillaumePrata commented:
> how does this improve security of said department theres like 20 welding tools free on the map roundstart

One day we will have tools be medium sized so not everyone and their mom have a full break in toolset inside a box, inside their bag, every single round.

For now, making it be slight slower is the best we can do.

Fikou commented:
why is it even movable if we removed the reason its moved

RedBaronFlyer committed:
(hopefully) fixes CL Suite check failure

I am stupid and deleted the bottom empty row and this was causing issues for CL Suite / Run Linters. Thank you Guillaume for pointing that out.

GPeckman commented:
> how does this improve security of said department theres like 20 welding tools free on the map roundstart

In fairness, welding masks are a lot less common and also harder to carry due to being normal-sized.

Mothblocks commented on code/modules/mining/machine_redemption.dm:
```diff
@@ -423,3 +431,88 @@
+// welding code
+
+/obj/machinery/mineral/ore_redemption/can_be_unfasten_wrench(mob/user, silent)
+	if(welded_down)
```
Check `silent`

Mothblocks commented on code/modules/mining/machine_redemption.dm:
```diff
@@ -423,3 +431,88 @@
+
+/obj/machinery/mineral/ore_redemption/can_be_unfasten_wrench(mob/user, silent)
+	if(welded_down)
+		to_chat(user, span_warning("[src] is welded to the floor!"))
```
Balloon alert

Mothblocks commented on code/modules/mining/machine_redemption.dm:
```diff
@@ -423,3 +431,88 @@
+	if(!anchored && welded_down) //make sure they're keep in sync in case it was forcibly unanchored by badmins or by a megafauna.
+		welded_down = FALSE
+	can_atmos_pass = anchorvalue ? ATMOS_PASS_NO : ATMOS_PASS_YES
+	air_update_turf(TRUE, anchorvalue)
```
Named parameter for TRUE

Mothblocks commented on code/modules/mining/machine_redemption.dm:
```diff
@@ -423,3 +431,88 @@
+	if(welded_down)
+		if(!tool.tool_start_check(user, amount=2))
+			return TRUE
+		user.visible_message(
+			span_notice("[user.name] starts to cut the [name] free from the floor."),
+			span_notice("You start to cut [src] free from the floor..."),
+			span_hear("You hear welding."),
+		)
```
balloon_alert_to_viewer

Mothblocks commented on code/modules/mining/machine_redemption.dm:
```diff
@@ -423,3 +431,88 @@
+		if(!tool.use_tool(src, user, delay=100, volume=100))
+			return FALSE
+		welded_down = FALSE
+		to_chat(user, span_notice("You cut [src] free from the floor."))
```
balloon_alert

Mothblocks commented on code/modules/mining/machine_redemption.dm:
```diff
@@ -423,3 +431,88 @@
+		to_chat(user, span_notice("You cut [src] free from the floor."))
+		return TRUE
+	if(!anchored)
+		to_chat(user, span_warning("[src] needs to be wrenched to the floor!"))
```
you must be noticing a pattern here. here and everywhere else

Mothblocks commented on code/modules/mining/machine_redemption.dm:
```diff
@@ -423,3 +431,88 @@
+		span_notice("You start to weld [src] to the floor..."),
+		span_hear("You hear welding."),
+	)
+	if(!tool.use_tool(src, user, delay=100, volume=100))
```
```suggestion
	if(!tool.use_tool(src, user, delay = 10 SECONDS, volume=100))
```

Mothblocks commented on code/modules/mining/machine_redemption.dm:
```diff
@@ -423,3 +431,88 @@
+		span_hear("You hear welding."),
+	)
+	if(!tool.use_tool(src, user, delay=100, volume=100))
+		balloon_alert(user, "cancelled!")
```
don't need this balloon alert because the do_after already has a visual tell

Mothblocks commented on code/modules/mining/machine_redemption.dm:
```diff
@@ -423,3 +431,88 @@
+		return NONE
+
+	var/tool_tip_set = FALSE
+	if(held_item.tool_behaviour == TOOL_WELDER && !istype(src, /obj/machinery/mineral/ore_redemption))
```
When would this istype ever pass? This doesn't make any sense

Loafin34 commented:
@GuillaumePrata what about pocket crowbars/screwdrivers? why would these be normal-sized and only fit in toolbelts? the welding tool and wrench and wirecutters make some sense, but pocket crowbars are.. pocketable

mc-oofert commented:
> why is it even movable if we removed the reason its moved

also this is a fair point why not just make it not unanchorable

if someone deconstructs it to tide you can call them an antag because thats an objective and legally lynch them

GuillaumePrata commented:
> @GuillaumePrata what about pocket crowbars/screwdrivers? why would these be normal-sized and only fit in toolbelts? the welding tool and wrench and wirecutters make some sense, but pocket crowbars are.. pocketable

We simply... remove the pocket crowbars from the game and just keep the medium sized ones...
Screwdrivers are whatever and so are multitools and wirecutters, not every tool needs to be medium sized to end the break in kit in a box which costs a whooping 1/49th of a player backpack per tool.

I went on a tangent so to go back on topic, the PR that made Rwalls be "slightly" harder to break in was one of the best changes in the game IMO, meaningful break ins make the round drama a lot more interesting, and the game will benefit from a slightly harder to break in cargo.

zxaber commented:
Can it still be deconstructed with a screwdriver and crowbar, or does being welded prevent that?

Does anyone actually use the ORM to dispense mats? Does it even need to be public facing anymore?

RedBaronFlyer committed:
Changes done as per Mothblock's request

Thank you Coding Channel people I am clueless.

zxaber commented:
>It can still be deconstructed with a screwdriver and a crowbar, though you would also need to wrench the machine frame and move it out of the way

I would bet cash money that this will lead to it being deconstructed most rounds, and now you'll have to deal with tiders in cargo *and* rebuilding/re-syncing the ORM.

Can we add a check in the deconstruction proc to fail if it's welded?

RedBaronFlyer committed:
Bug Fix for Balloon alerts, and an attempt to make balloon alerts better comply with modern balloon alert text formatting.

RedBaronFlyer commented:
> > It can still be deconstructed with a screwdriver and a crowbar, though you would also need to wrench the machine frame and move it out of the way
> 
> I would bet cash money that this will lead to it being deconstructed most rounds, and now you'll have to deal with tiders in cargo _and_ rebuilding/re-syncing the ORM.
> 
> Can we add a check in the deconstruction proc to fail if it's welded?

I'll see what I can do.

RedBaronFlyer commented:
Alright, so this vastly went over my head in no time flat. So the plan was just to copy over the code that the smartfridge uses so that cargo didn't get tided into every round. I wasn't expecting to need to update things to comply with modern TG coding standards, though I completely understand why this would be requested of me. There is also the aforementioned issue of tiders breaking down the ORM as opposed to welding it. I simply lack the ability to get this working in a way that I would find satisfactory and in a manner that would be acceptable to TG coding standards.

I fully get that, and I am not blaming anyone but myself for what I got myself into.

Anyone else can pick this up in the future if they want.
